tower-retry/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+#[macro_use]
+extern crate futures;
+extern crate tower_service;
+
+use futures::{Async, Future, Poll};
+use tower_service::Service;
+
+#[derive(Clone, Debug)]
+pub struct Retry<P, S> {
+    policy: P,
+    service: S,
+}
+
+#[derive(Debug)]
+pub struct ResponseFuture<P: Policy<S::Request, S::Response, S::Error>, S: Service> {
+    request: Option<S::Request>,
+    retry: Retry<P, S>,
+    state: State<S::Future, P::Future, S::Response, S::Error>,
+}
+
+#[derive(Debug)]
+enum State<F, P, R, E> {
+    /// Polling the future from `Service::call`
+    Called(F),
+    /// Polling the future from `Policy::retry`
+    Checking(P, Option<Result<R, E>>),
+    /// Polling `Service::poll_ready` after `Checking` was OK.
+    Retrying,
+}
+
+pub trait Policy<Req, Res, E>: Sized {
```

Comment:
Why are these not associated types? I have some idea, but it would be helpful to explain it out.

If ATs don't work, `Policy` could be generic over `S: Service` to keep it a bit more succinct?

Also, why the `Sized` bound?

Reply:
I started with generic parameters since they're "input" types, not "output" types. That's as far as my thinking on the matter has gone.

Reply:
The `Sized` bound was needed to allow `type Future = Future<Item = Self>`.